### ISSUE TITLE
OpenID Connect / How to setup Identity Propagation for RESTful Resources 

### ISSUE THREAD

**Issue by claudefalbriard** [mitreid-connect/OpenID-Connect-Java-Spring-Server#1172]
Dear colleagues, 

 I'd like to implement **OpenID Connect Identity Propagation** feature for RESTful resources that are located in the chain of services beyond the OpenID Connect login authentication.  Could you describe how to setup the RESTful servers, or by use of OAuth2 tokens, or by passage of the the self-contained ID tokens obtained from the OIDC APIs. 

My guess is that from a performance standpoint, the use of JWT would be the preferred method, as it avoids unnecessary look-ups of OIDC /intospect, or /userinfo function. It also relies on the JWK, a newer tool for key validation and cryptographic support.  
![identity_propagation_oidc](https://cloud.githubusercontent.com/assets/23198783/21767506/c4752ba8-d65b-11e6-990c-95e9ac4c4aae.png)
  

For more details and diagram, please check the report I recorded under the Stackoverflow forum: 
[](http://stackoverflow.com/questions/41485208/openid-connect-restful-identity-propagation-with-mitreid-client)

Thanks in advance for your advice and guidance. 

Regards, 
Claude

**Comment by jricher** [comment]
I would not recommend deploying what is in your diagram above. You should never pass your ID token to another application. The ID token is a single use assertion pointed to the client application, which is whatever application you're logging in to. If you need to access a back end service for an app, that's exactly what OAuth was designed to solve. You can use the introspection endpoint (recommended for liveness of information like token revocation status) or JWT formatted tokens  (recommended for less network traffic) to convey information about the current resource owner to the backend API. It's a tradeoff decision on which way you want to go, and sometimes you can do both (it's legal to introspect a JWT instead of parsing it, and that's exactly what our server allows by default FWIW). 

It's also important that your backend services don't confuse the incoming access token as a "login" by  a user and treat the request that way. Access tokens are not logins, they're authorization delegations and should be treated appropriately.

So in short with the MITREid Connect project, you already get JWT formatted tokens and access to an introspection endpoint, so you can already use the server to log in to an OIDC client application and that client application can call backend APIs using the OAuth access token, which the APIs can introspect and/or parse as they see fit.

**Comment by claudefalbriard** [comment]
Thanks for the fast reply and direction. I've tested the /introspect function on our IdP server (Tivoli FIM) and got a positive result from the request returning the following data from my request:  
{"exp":1484053138,"realmName":"mga","sub":"myname@red.com","scope":"openid","grant_type":"refresh_token","active":true,"token_type":"Bearer","client_id":"abc123","iat":1484045944}

Sending this request requires a client_id and client_secret, so I expect that the introspect will be executed by the resource server using the MITREID client. Now there remains a doubt how to trigger this process in a typical REST API execution flow. 

Our web server runs Oauth2 & OIDC login authentication that acts like a portal and dispatches the API call (REST consumer) to a REST resource (REST provider), usually by doing HTTPS GET or POST operations. What are the mandatory arguments required at the HTTP passage to invoke the /introspect operation at the resource server side?  Is it just the passage the of the Oauth2 access token like: "Authorization: Bearer <_my_access_token_>", or some additional arguments like: scope,  grant_type=client_credential, or others? For reference, Is there a syntax sample available in CURL?

**Comment by jricher** [comment]
You need to register a separate "client" in the server for the protected resource and turn on "allow token introspection" for it, that gives you the ID and secret for your resource. 

You seem to already have the call to introspection working: The syntax is a simple POST to the introspection endpoint with the token in the body and the resource's ID and secret as an HTTP auth header. Perhaps you should read the introspection specification: https://tools.ietf.org/html/rfc7662
